src/geometry/detail/VGNavCollection.hh

```diff
@@ -0,0 +1,182 @@
+// See the top-level COPYRIGHT file for details.
+// SPDX-License-Identifier: (Apache-2.0 OR MIT)
+//---------------------------------------------------------------------------//
+//! \file VGNavCollection.hh
+//---------------------------------------------------------------------------//
+#pragma once
+
+#ifndef __CUDA_ARCH__
+#    include <memory>
+#endif
+#include <VecGeom/navigation/NavigationState.h>
+#include "base/Assert.hh"
+#include "base/OpaqueId.hh"
+#include "base/Types.hh"
+
+namespace vecgeom
+{
+#ifndef __NVCC__
+inline
+#endif
+    namespace cxx
+{
+class NavStatePool;
+} // namespace cxx
+} // namespace vecgeom
+
+// CUDA declarations aren't reachable from C++ host-compiled code, and the
+// reverse, so we must fudge by forward-declaring here as though
+// NavigationState is an untemplated class, not a type alias, for each unused
+// memspace.
+namespace vecgeom
+{
+#ifdef __NVCC__
+namespace cxx
+#else
+namespace cuda
+#endif
+{
+class NavigationState;
+}
+} // namespace vecgeom
+
+namespace celeritas
+{
+namespace detail
+{
+template<Ownership W, MemSpace M>
+struct VGNavCollection;
+
+// Declare specializations because each one is different and may not be
+// defined based on active compiler/phase
+template<>
+struct VGNavCollection<Ownership::value, MemSpace::host>;
+template<>
+struct VGNavCollection<Ownership::value, MemSpace::device>;
+template<>
+struct VGNavCollection<Ownership::reference, MemSpace::host>;
+template<>
+struct VGNavCollection<Ownership::reference, MemSpace::device>;
+
+#ifndef __CUDA_ARCH__
+template<>
+struct VGNavCollection<Ownership::value, MemSpace::host>
+{
+    using NavState = vecgeom::cxx::NavigationState;
+
+    std::unique_ptr<NavState> nav_state;
+
+    void resize(int max_depth, int size);
+
+    explicit operator bool() const { return static_cast<bool>(nav_state); }
+};
+
+template<>
+struct VGNavCollection<Ownership::reference, MemSpace::host>
+{
+    using NavState = vecgeom::cxx::NavigationState;
+
+    NavState* ptr = nullptr;
+
+    VGNavCollection& operator=(const VGNavCollection&) = default;
+
+    void operator=(VGNavCollection<Ownership::value, MemSpace::host>& other);
+
+    NavState& at(int, ThreadId id) const
+    {
+        CELER_EXPECT(ptr);
+        CELER_EXPECT(id.get() == 0);
+        return *ptr;
+    }
+
+    explicit operator bool() const { return static_cast<bool>(ptr); }
+};
+
+//---------------------------------------------------------------------------//
+/*!
+ * Delete a VecGeom pool.
+ *
+ * Due to vecgeom macros, the definition of this function can only be compiled
+ * from a .cc file.
+ */
+struct NavStatePoolDeleter
+{
+    using arg_type = vecgeom::cxx::NavStatePool*;
+    void operator()(arg_type) const;
+};
+
+//---------------------------------------------------------------------------//
+/*!
+ * Manage a pool of device-side geometry states.
+ *
+ * Construction of the navstatepool has to be in a host compliation unit due to
```

Comment:
typos: _compilation_ and `NavStatePool`.
(I could not figure out how to make the changes as a suggestion), as I've seen done by @sethrj)